File: backend/app/schemas/learning.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class DoodlePredictionRequest(BaseModel):
    child_id: UUID
    assignment_id: UUID
    image_data_url: str = Field(min_length=20)
    target_class: str | None = Field(default=None, max_length=60)
    top_k: int = Field(default=3, ge=1, le=10)


class DoodlePredictionResponse(BaseModel):
    target_class: str | None
    predicted_class: str
    english: str
    vi: str
    phonetic: str | None
    confidence: float
    is_correct: bool
    top_predictions: list[dict[str, Any]]


class SubmissionCreate(BaseModel):
    child_id: UUID
    assignment_id: UUID
    target_class: str | None = Field(default=None, max_length=60)
    predicted_class: str = Field(max_length=60)
    confidence: float = Field(ge=0, le=1)
    is_correct: bool | None = None
    top_predictions: list[dict[str, Any]] | None = None
    canvas_image_data_url: str | None = Field(default=None, min_length=20)
    speech_transcript: str | None = Field(default=None, max_length=1000)
    speech_passed: bool = False


class SubmissionPublic(BaseModel):
    id: UUID
    assignment_id: UUID
    child_id: UUID
    submission_type: str
    target_class: str | None
    predicted_class: str | None
    confidence: float | None
    is_correct: bool
    top_predictions: list[dict[str, Any]] | None
    canvas_image_url: str | None
    speech_transcript: str | None
    speech_passed: bool
    stars_earned: int
    coins_earned: int
    created_at: datetime

File: frontend/src/lib/learning.ts
import { apiRequest } from "./api";

export type DoodlePrediction = {
  target_class: string | null;
  predicted_class: string;
  english: string;
  vi: string;
  phonetic: string | null;
  confidence: number;
  is_correct: boolean;
  top_predictions: Array<{
    class_key: string;
    english: string;
    vi: string;
    phonetic: string | null;
    confidence: number;
  }>;
};

export type SubmissionResult = {
  id: string;
  assignment_id: string;
  child_id: string;
  target_class: string | null;
  predicted_class: string;
  confidence: number;
  is_correct: boolean;
  top_predictions: DoodlePrediction["top_predictions"] | null;
  canvas_image_url: string | null;
  speech_transcript: string | null;
  speech_passed: boolean;
  stars_earned: number;
  coins_earned: number;
  created_at: string;
};

export function predictDoodle(
  token: string,
  payload: {
    child_id: string;
    assignment_id: string;
    image_data_url: string;
    target_class?: string | null;
    top_k?: number;
  },
): Promise<DoodlePrediction> {
  return apiRequest<DoodlePrediction>("/ai/predict", {
    method: "POST",
    token,
    body: JSON.stringify(payload),
  });
}

export function createSubmission(
  token: string,
  payload: {
    child_id: string;
    assignment_id: string;
    target_class?: string | null;
    predicted_class: string;
    confidence: number;
    is_correct?: boolean;
    top_predictions?: DoodlePrediction["top_predictions"];
    canvas_image_data_url?: string | null;
    speech_transcript?: string | null;
    speech_passed: boolean;
  },
): Promise<SubmissionResult> {
  return apiRequest<SubmissionResult>("/submissions", {
    method: "POST",
    token,
    body: JSON.stringify(payload),
  });
}
